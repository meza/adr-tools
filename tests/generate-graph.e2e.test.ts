/* eslint-disable no-sync */
import { describe, it, expect, afterAll, beforeAll, vi, afterEach } from 'vitest';
import * as childProcess from 'child_process';
import * as path from 'path';
import * as fs from 'fs/promises';
import * as os from 'os';
import { realpathSync } from 'node:fs';

describe('Generating Graphs', () => {
  const adr = path.resolve(path.dirname(__filename), '../src/index.ts');
  const command = `npx tsx ${adr}`;
  let workDir: string;

  beforeAll(async () => {
    // @ts-ignore
    process.env.ADR_DATE = '1992-01-12';
    workDir = path.resolve(realpathSync(await fs.mkdtemp(path.join(os.tmpdir(), 'adr-'))));
    childProcess.execSync(`${command} init`, { cwd: workDir });
    childProcess.execSync(`${command} new An idea that seems good at the time`, { cwd: workDir });
    childProcess.execSync(`${command} new -s 2 A better idea`, { cwd: workDir });
    childProcess.execSync(`${command} new This will work`, { cwd: workDir });
    childProcess.execSync(`${command} new -s 3 The end`, { cwd: workDir });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  afterAll(() => {
    childProcess.execSync(`rimraf ${workDir}`);
  });

  it('should generate a graph', async () => {
    const child = childProcess.execSync(`${command} generate graph`, { cwd: workDir });
    const childContent = child.toString().trim();

    expect(childContent).toMatchSnapshot();
  });

  it('should generate a graph with specified route and extension ', async () => {
    const child = childProcess.execSync(`${command} generate graph -p http://example.com/ -e .xxx`, { cwd: workDir });
    const childContent = child.toString().trim();

    expect(childContent).toMatchSnapshot();
  });
});
