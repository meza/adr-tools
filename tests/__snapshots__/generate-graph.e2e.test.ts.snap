// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generating Graphs > should generate a graph 1`] = `
"digraph {
  node [shape=plaintext];
  subgraph {
    _1 [label="1. Record architecture decisions"; URL="0001-record-architecture-decisions.html"];
    _2 [label="2. An idea that seems good at the time"; URL="0002-an-idea-that-seems-good-at-the-time.html"];
    _1 -> _2 [style="dotted", weight=1];
    _3 [label="3. A better idea"; URL="0003-a-better-idea.html"];
    _2 -> _3 [style="dotted", weight=1];
    _4 [label="4. This will work"; URL="0004-this-will-work.html"];
    _3 -> _4 [style="dotted", weight=1];
    _5 [label="5. The end"; URL="0005-the-end.html"];
    _4 -> _5 [style="dotted", weight=1];
  }
  _3 -> _2 [label="Supersedes", weight=0]
  _5 -> _3 [label="Supersedes", weight=0]
}"
`;

exports[`Generating Graphs > should generate a graph with specified route and extension  1`] = `
"digraph {
  node [shape=plaintext];
  subgraph {
    _1 [label="1. Record architecture decisions"; URL="http://example.com/0001-record-architecture-decisions.xxx"];
    _2 [label="2. An idea that seems good at the time"; URL="http://example.com/0002-an-idea-that-seems-good-at-the-time.xxx"];
    _1 -> _2 [style="dotted", weight=1];
    _3 [label="3. A better idea"; URL="http://example.com/0003-a-better-idea.xxx"];
    _2 -> _3 [style="dotted", weight=1];
    _4 [label="4. This will work"; URL="http://example.com/0004-this-will-work.xxx"];
    _3 -> _4 [style="dotted", weight=1];
    _5 [label="5. The end"; URL="http://example.com/0005-the-end.xxx"];
    _4 -> _5 [style="dotted", weight=1];
  }
  _3 -> _2 [label="Supersedes", weight=0]
  _5 -> _3 [label="Supersedes", weight=0]
}"
`;
